src/Disks/VolumeJBOD.h

```diff
@@ -63,8 +64,35 @@ class VolumeJBOD : public IVolume
     bool are_merges_avoided = true;
 
 private:
-    /// Index of last used disk.
+    struct DiskWithSize
+    {
+        DiskPtr disk;
+        uint64_t free_size = 0;
+
+        DiskWithSize(DiskPtr disk_)
+            : disk(disk_)
+            , free_size(disk->getUnreservedSpace())
```

Comment:
I'm not aware if there is a way to modify unreserved space from somewhere else? You mention it could be a problem. So I think to avoid problem of `free_space` not being equal to `getUnreservedSpace()` for long time we can update `free_size` with new value of unreserved space between every pop() and push(). It can still be out of sync while it is inside priority queue and it is okay - we use only pop()ed items. We can even check if it is out of sync after pop() and before using it, and if it is, we can update key and push it back and then retry with another pop(). Probably retries doesn't worth if these out-of-sync are rare, but just updating key before pushing will be good imo.

Also with this logic you are not relying on idea that reserve will reduce `free_space` by exactly `bytes`, which is implementation detail of DIsk itself. You can just call reserve and then ask unreserved space.

Reply:
>I'm not aware if there is a way to modify unreserved space from somewhere else?

```
$ gg 'disk.*reserve('
src/Disks/SingleDiskVolume.h:        return disks[0]->reserve(bytes);
src/Disks/StoragePolicy.cpp:    auto reservation = max_disk->reserve(0);
src/Disks/VolumeJBOD.cpp:                auto reservation = disks[index]->reserve(bytes);
src/Disks/VolumeJBOD.cpp:            ReservationPtr reservation = disk.reserve(bytes);
src/Disks/VolumeJBOD.h:            ReservationPtr reservation = disk->reserve(bytes);
src/Storages/MergeTree/MergeTreeData.cpp:        reservation = selected_disk->reserve(expected_size);
src/Storages/System/StorageSystemStoragePolicies.cpp:            disks.reserve(volumes[i]->getDisks().size());
```

> You mention it could be a problem.

Yes, but this should not be major (plus some allocations can be done not by clickhouse).

>So I think to avoid problem of free_space not being equal to getUnreservedSpace() for long time we can update free_size with new value of unreserved space between every pop() and push(). It can still be out of sync while it is inside priority queue and it is okay - we use only pop()ed items. We can even check if it is out of sync after pop() and before using it, and if it is, we can update key and push it back and then retry with another pop(). Probably retries doesn't worth if these out-of-sync are rare, but just updating key before pushing will be good imo.

I thought about this, but this will requires extra statfs().
I want to fix this separately to change the reservation interface, but OK, I did it inside this patch set.